cranelift/codegen/src/isa/x64/inst/args.rs

```diff
@@ -107,7 +108,7 @@ macro_rules! newtype_of_reg {
         $(
             /// A newtype wrapper around `RegMem` for general-purpose registers.
             #[derive(Clone, Debug)]
-            pub struct $newtype_reg_mem(RegMem);
+            pub struct $newtype_reg_mem(pub RegMem);
```

Comment:
Hmm, yeah, in principle I agree, but I couldn't find a way to make the pattern-matches see through these. Without the ability to pattern-match `Inst::Inst { src: GprMem(RegMem::Mem { ... }), .. }` vs. `Inst::Inst { src: GprMem(RegMem::Reg { ... }), .. }`, the code in `isa::x64::pcc` either becomes nightmarish or just impossible. I think we need to find a way to preserve that somehow...

Reply:
The right long-term answer -- both typesafe and compatible with pattern-matching -- would be to directly have `enum GprMem { Gpr { ... }, Mem { ... } }`; that way the type still names the invariant, but we also can pattern-match on the interior. I don't want to couple that refactor to this work though, and unfortunately I don't see any trait impls that will let pattern-matching work otherwise; I would love to be wrong though if you know a way?

Reply:
Instead of only matching when the `{Gpr,Xmm}Mem` wraps a `RegMem::Mem`, we can always match and then call `operand.as_ref()` or `operand.into()` to get the underlying `RegMem`, and finally pass that into a helper that early exits when given a `RegMem::Reg`. Essentially pushing the `::Reg` vs `::Mem` filtering down one layer.

Reply:
Right, I get that this is a workaround, but I'm saying that the code becomes nightmarish: we really do need the ability to match on the Inst overall with specific kinds of operands...

Reply:
(Spoke offline with fitzgen a bit; I'm going to try to refactor in this direction anyway and see how bad it really is. Long-term it would still be nice to get to separate enums to allow native deep sub-pattern-matching!)